auth/TESLA/Receiver.m

```diff
@@ -0,0 +1,135 @@
+classdef Receiver < handle
+    % RECIEVER Summary of this class goes here
+    %   Detailed explanation goes here
+
+    properties
+        key_chain_function = @HashingWrappers.sha_256
+        mac_signing_function = @HashingWrappers.hmac_sha_256
+        message_table
+        reciever_key_chain = cell(1, 100)
+        next_key_chain = cell(1, 100)
+        verified_block = cell (1, 5)
+        starting_hash
+        mt50_table
+        missing_mt50 = {}
+    end
+
+    methods
+
+        function obj = Receiver(hp, hp2)
+            obj.reciever_key_chain{1} = hp;
+            obj.next_key_chain{1} = hp2;
+            obj.message_table = containers.Map('KeyType', 'double', 'ValueType', 'any');
+            obj.mt50_table = containers.Map('KeyType', 'double', 'ValueType', 'any');
+        end
+
+        function add_message(obj, message, time)
+            to_add = cell(1, 2);
+            to_add{1} = message;
+            to_add{2} = false;
+
+            obj.message_table(time) = to_add;
+
+            if isa(message, 'MT50')
+                if time == 6
```

Comment:
The logic should not depend on time here. If time = 6, obj.verify_mt50 will hash to a obj.reciever_key_chain{1} or obj.next_key_chain{1}; otherwise, it will hash down to obj.reciever_key_chain{2...?}

Reply:
your right but i also call the hmac verification in the else part of the statement and there will be no hmacs to verify for hp 2